File: org.summer.view.widget/src/org/summer/view/widget/TemplateBindingExtension.java
package org.summer.view.widget;

import org.summer.view.widget.data.IValueConverter;
import org.summer.view.widget.markup.MarkupExtension;

/// <summary>
///  Class for Xaml markup extension for TemplateBindings that
///  can be set on the nodes of the Template VisualTree.
/// </summary> 
//[TypeConverter(typeof(TemplateBindingExtensionConverter))]
//[MarkupExtensionReturnType(typeof(Object))] 
public class TemplateBindingExtension extends MarkupExtension 
{
    /// <summary> 
    ///  Constructor that takes no parameters
    /// </summary>
    public TemplateBindingExtension()
    { 
    }

    /// <summary> 
    ///  Constructor that takes the resource key that this is a static reference to.
    /// </summary> 
    public TemplateBindingExtension(
        DependencyProperty property)
    {
        if (property != null) 
        {
            _property = property; 
        } 
        else
        { 
            throw new ArgumentNullException("property");
        }
    }

    /// <summary>
    ///  Return an Object that should be set on the targetObject's targetProperty 
    ///  for this markup extension.  For TemplateBindingExtension, this is the Object found in 
    ///  a resource dictionary in the current parent chain that is keyed by ResourceKey
    /// </summary> 
    /// <param name="serviceProvider">ServiceProvider that can be queried for services.</param>
    /// <returns>
    ///  The Object to set on this property.
    /// </returns> 
    public /*override*/ Object ProvideValue(IServiceProvider serviceProvider)
    { 
        if (Property == null) 
        {
            throw new InvalidOperationException(/*SR.Get(SRID.MarkupExtensionProperty)*/); 
        }

        return new TemplateBindingExpression(this);
    } 

    /// <summary> 
    ///     Property we are binding to 
    /// </summary>
//    [ConstructorArgument("property")] 
    public DependencyProperty Property
    {
        get { return _property; }
        set 
        {
            if (value == null) 
            { 
                throw new ArgumentNullException("value");
            } 
            _property = value;
        }
    }

    /// <summary>
    ///     ValueConverter to interpose between the source and target properties 
    /// </summary> 
//    [DefaultValue(null)]
    public IValueConverter Converter 
    {
        get { return _converter; }
        set
        { 
            if (value == null)
            { 
                throw new ArgumentNullException("value"); 
            }
            _converter = value; 
        }
    }

    /// <summary> 
    ///     ConverterParameter we are binding to
    /// </summary> 
//    [DefaultValue(null)] 
    public Object ConverterParameter
    { 
        get { return _parameter; }
        set { _parameter = value; }
    }

    private DependencyProperty _property;
    private IValueConverter _converter; 
    private Object _parameter; 
}

File: summer/WebContent/js/system/windows/TemplateBindingExtension.js
/**
 * TemplateBindingExtension
 */

define(["dojo/_base/declare", "system/Type", "markup/MarkupExtension", "windows/TemplateBindingExpression"],
		function(declare, Type, MarkupExtension, TemplateBindingExpression){
	var TemplateBindingExtension = declare("TemplateBindingExtension", MarkupExtension,{
		constructor:function(/*DependencyProperty*/ property)
        {
			if(property === undefined){
				property = null;
			}
			
            if (property != null) {
                this._property = property; 
            } 
        },
        
        ///  Return an object that should be set on the targetObject's targetProperty 
        ///  for this markup extension.  For TemplateBindingExtension, this is the object found in 
        ///  a resource dictionary in the current parent chain that is keyed by ResourceKey
        /// <param name="serviceProvider">ServiceProvider that can be queried for services.</param>
        /// <returns>
        ///  The object to set on this property.
        /// </returns> 
//        public override object 
        ProvideValue:function(/*IServiceProvider*/ serviceProvider)
        { 
            if (this.Property == null){
                throw new Error('InvalidOperationException(SR.Get(SRID.MarkupExtensionProperty)'); 
            }

            return new TemplateBindingExpression(this);
        } 
        
	});
	
	Object.defineProperties(TemplateBindingExtension.prototype,{
        ///     Property we are binding to 
//        public DependencyProperty 
        Property:
        {
            get:function() { return this._property; },
            set:function(value) 
            {
                if (value == null) 
                { 
                    throw new ArgumentNullException("value");
                } 
                this._property = value;
            }
        },
 
        ///     ValueConverter to interpose between the source and target properties 
//        public IValueConverter 
        Converter:
        {
            get:function() { return this._converter; },
            set:function(value)
            { 
                if (value == null)
                { 
                    throw new ArgumentNullException("value"); 
                }
                this._converter = value; 
            }
        },

        ///     ConverterParameter we are binding to
//        public object 
        ConverterParameter:
        { 
            get:function() { return this._parameter; },
            set:function(value) { this._parameter = value; }
        }
	});
	
	TemplateBindingExtension.Type = new Type("TemplateBindingExtension", TemplateBindingExtension, [MarkupExtension.Type]);
	return TemplateBindingExtension;
});
 


 
 
//        private DependencyProperty _property;
//        private IValueConverter _converter; 
//        private object _parameter; 
//    }
//} 

